### PR TITLE
extract code-gen utilities from closures

### PR DESCRIPTION
**Pull request by esc**
This is just a refactoring of codegen utilities for the typed list. Some
of these codegen functions will be used later on when unboxing into a
typed-list. No functional changes were made, so no tests needed
updating.

### CODE REVIEW

File: numba/listobject.py [hunk 1]
```diff
@@ -86,13 +86,23 @@ class ListStatus(IntEnum):
...
+        self.context = context
+
+    def __call__(self, builder, status, msg):
+        ok_status = status.type(int(ListStatus.LIST_OK))
+        with builder.if_then(builder.icmp_signed('!=', status, ok_status),
+                             likely=True):
+            self.context.call_conv.return_user_exc(
+                builder, RuntimeError, (msg,))
+
+
```
Comment (stuartarchibald) [inline numba/listobject.py]: flake8 complains about too many blank lines here now.

Comment (esc) [reply numba/listobject.py]: yeah, not sure how that made it through my `pre-commit` hooks.

Comment (esc) [reply numba/listobject.py]: fixed and pushed

**Comment by esc** [comment]
@sklam thanks for merging!

**Pull request merged**